exchange.c

```diff
@@ -103,77 +103,83 @@ int domain_exchange(int (*layoutfunc)(int p)) {
     return failure;
 }
 
+#define NSP 3
+
 static int domain_exchange_once(int (*layoutfunc)(int p), int* toGo, int * toGoSph, int * toGoBh, int *toGet, int *toGetSph, int *toGetBh)
 {
-    int count_togo = 0, count_togo_sph = 0, count_togo_bh = 0, 
-        count_get = 0, count_get_sph = 0, count_get_bh = 0;
-
-    int i, target;
+    int i, j, target;
     struct particle_data *partBuf;
     struct sph_particle_data *sphBuf;
     struct bh_particle_data *bhBuf;
 
-    int * count = ta_malloc("var", int, NTask);
-    int * count_sph = ta_malloc("var", int, NTask);
-    int * count_bh = ta_malloc("var", int, NTask);
-    int * offset = ta_malloc("var", int, NTask);
-    int * offset_sph = ta_malloc("var", int, NTask);
-    int * offset_bh = ta_malloc("var", int, NTask);
-
-    int * count_recv = ta_malloc("var", int, NTask);
-    int * count_recv_sph = ta_malloc("var", int, NTask);
-    int * count_recv_bh = ta_malloc("var", int, NTask);
-    int * offset_recv = ta_malloc("var", int, NTask);
-    int * offset_recv_sph = ta_malloc("var", int, NTask);
-    int * offset_recv_bh = ta_malloc("var", int, NTask);
-
-    for(i = 1, offset_sph[0] = 0; i < NTask; i++)
-        offset_sph[i] = offset_sph[i - 1] + toGoSph[i - 1];
-
-    for(i = 1, offset_bh[0] = 0; i < NTask; i++)
-        offset_bh[i] = offset_bh[i - 1] + toGoBh[i - 1];
-
-    offset[0] = 0;
+    int *ctmem = mymalloc("cts", 4*NSP*NTask*sizeof(int));
```

Comment:
does it hurt to keep NSP at 6 and use the ptype to index these arrays?

Reply:
No it is more complicated because we are dealing with the base as well.

Reply:
Yes, because 0 is "total": we would need NSP = 7 and then it gets complex.